src/puppetlabs/puppetdb/queue.clj

```diff
@@ -0,0 +1,199 @@
+  ([valid-commands]
+   ;; NOTE: changes here may affect the DLO, e.g. it currently assumes
+   ;; the trailing .json.
+   (let [rx (metadata-rx valid-commands)]
+     (fn [s]
+       (when-let [[_ stamp command version certname] (re-matches rx s)]
+         (and certname
+              {:received (-> stamp
+                             Long/parseLong
+                             tcoerce/from-long
+                             kitchensink/timestamp)
+               :version (Long/parseLong version)
+               :command command
+               :certname certname}))))))
+
+(def parse-metadata (metadata-parser))
+
+(def parse-cmd-filename
+  (let [parse-metadata (metadata-parser (cons "unknown" metadata-command-names))]
+    (fn [s]
+      (let [rx #"([0-9]+)-(.*)"]
+        (when-let [[_ id qmeta] (re-matches rx s)]
+          (parse-metadata qmeta))))))
+
+(defrecord CommandRef [id command version certname received callback annotations delete?])
+
+(defn cmdref->entry [{:keys [id command version certname received]}]
+  (stock/entry id (metadata-str received command version certname)))
+
+(defn entry->cmdref [entry]
+  (let [{:keys [received command version certname]} (-> entry
+                                                        stock/entry-meta
+                                                        parse-metadata)]
+    (map->CommandRef {:id (stock/entry-id entry)
+                      :command command
+                      :version version
+                      :certname certname
+                      :received received
+                      :callback identity
+                      :annotations {:id (stock/entry-id entry)
+                                    :received received}})))
+
+(defn cmdref->cmd [q cmdref]
+  (let [entry (cmdref->entry cmdref)]
+    (with-open [command-stream (stock/stream q entry)]
+      (assoc cmdref
+             :payload (stream->json command-stream)
+             :entry entry))))
+
+(defn cons-attempt [cmdref exception]
+  (update cmdref :attempts conj {:exception exception
+                                 :time (kitchensink/timestamp)}))
+
+(defn store-command
+  ([q command version certname command-stream]
+   (store-command q command version certname command-stream identity))
+  ([q command version certname command-stream command-callback]
+   (let [current-time (time/now)
+         entry (stock/store q
+                            command-stream
+                            (metadata-str current-time command version certname))]
+     (map->CommandRef {:id (stock/entry-id entry)
+                       :command command
+                       :version version
+                       :certname certname
+                       :callback command-callback
+                       :received (kitchensink/timestamp current-time)
+                       :annotations {:id (stock/entry-id entry)
+                                     :received (kitchensink/timestamp current-time)}}))))
```

Comment:
I'll write up a ticket for the id/received stuff. I kept it like this because everything in command.clj was looking in annotations. I think a top level key is better, so new code is looking there.

For command strings, this is being fixed in one of @aperiodic's PRs. We'll be using the same string for all commands of that type, so it'll only cost a reference for each command.